lib/Optimizer/Transforms/UnitarySynthesis.cpp

```diff
@@ -0,0 +1,212 @@
+                   [&](auto element) { return element * inv_phase; });
+
+    auto abs_00 = std::abs(special_unitary[0]);
+    auto abs_01 = std::abs(special_unitary[1]);
+
+    if (abs_00 >= abs_01)
+      angles.beta = 2 * std::acos(abs_00);
+    else
+      angles.beta = 2 * std::asin(abs_01);
+
+    auto half_sum_alpha_gamma =
+        std::atan2(special_unitary[3].imag(), special_unitary[3].real());
+    auto half_diff_alpha_gamma =
+        std::atan2(special_unitary[2].imag(), special_unitary[2].real());
+
+    angles.alpha = half_sum_alpha_gamma + half_diff_alpha_gamma;
+    angles.gamma = half_sum_alpha_gamma - half_diff_alpha_gamma;
+  }
+
+  BasisZYZ(std::vector<std::complex<double>> vec) {
+    std::move(vec.begin(), vec.end(), matrix.begin());
+  }
+};
+
+class CustomUnitaryPattern
+    : public OpRewritePattern<quake::CustomUnitarySymbolOp> {
+
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(quake::CustomUnitarySymbolOp customOp,
+                                PatternRewriter &rewriter) const override {
+
+    auto parentModule = customOp->getParentOfType<ModuleOp>();
+
+    Location loc = customOp->getLoc();
+    auto targets = customOp.getTargets();
+    auto controls = customOp.getControls();
+
+    /// Get the global constant holding the concrete matrix corresponding to
+    /// this custom operation invocation
+    StringRef generatorName = customOp.getGenerator().getRootReference();
+    auto globalOp =
+        parentModule.lookupSymbol<cudaq::cc::GlobalOp>(generatorName);
+
+    /// The decomposed sequence of quantum operations are in a function
+    auto pair = generatorName.split(".rodata");
+    std::string funcName = pair.first.str() + ".kernel" + pair.second.str();
+
+    /// If the replacement function doesn't exist, create it here
+    if (!parentModule.lookupSymbol<LLVM::LLVMFuncOp>(funcName)) {
+
+      auto unitary = cudaq::readGlobalConstantArray(globalOp);
+
+      /// TODO: Expand the logic to decompose upto 4-qubit operations
+      if (unitary.size() != 4)
+        return customOp.emitError(
+            "Decomposition of only single qubit custom operations supported.");
+
+      /// Controls are handled via apply specialization
+      auto funcTy =
+          FunctionType::get(parentModule.getContext(), targets.getTypes(), {});
+      auto insPt = rewriter.saveInsertionPoint();
+      rewriter.setInsertionPointToStart(parentModule.getBody());
+      auto func = rewriter.create<func::FuncOp>(parentModule->getLoc(),
+                                                funcName, funcTy);
+      func.setPrivate();
+      rewriter.restoreInsertionPoint(insPt);
+
+      insPt = rewriter.saveInsertionPoint();
+      auto *block = func.addEntryBlock();
+      rewriter.setInsertionPointToStart(block);
+
+      /// Use Euler angle decomposition for single qubit operation
+      auto zyz = BasisZYZ(unitary);
+      zyz.decompose();
```

Comment:
Does it make sense to make the `decompose` call part of the constructor so that we don't need to call it manually?